Add systemd service to bitcoind in debian package

ctp-tsteenholdt opened the pull request:
On suggestion from @TheBlueMatt I have updated `contrib/debian` files to include a systemd service in the `bitcoind` build. Tested and working on Ubuntu 16.04 and 17.10.

This fixes Issue #12758

randolf commented:
@ctp-tsteenholdt Please remember to squash/rebase.

ctp-tsteenholdt commented:
@randolf That makes total sense. My apologies - looking into it right now...

laanwj commented:
We already have a `bitcoind.service` in `contrib/init/` - what is the difference with this one, can we avoid duplicating it?

ctp-tsteenholdt commented:
@randolf The pull request should look better now - please let me know if I'm missing something.

ctp-tsteenholdt commented:
@laanwj I'm completely with you on that question, but I don't think we can feasibly avoid multiple versions of a file like this. The different versions of the `bitcoind.service` in the `contrib/` tree have subtle differences to make them work best for their particular purpose. The differences reflect distro-specifics such as location of environment files, dependency services etc.

Also, building a package for Ubuntu or Debian requires a full copy of the `contrib/debian/` folder, including the `bitcoind.service` file, to allow the packaging helper `dh` to recognize, treat and install it correctly.

It think distro-specific files are probably better maintained within distro specific folders such as `contrib/debian/` or `contrib/rpm/`, as part of the overall package maintenance for that distro. This avoids possible breakage from a change made for a different distro too.

To underline that the `contrib/init/` version is still very relevant, it will be the goto file for people on platforms without a specific package or people who prefer to build from source. This file can be easily adapted to work well on most systems.

Please let me know what you think.

laanwj commented:
@ctp-tsteenholdt If it is intentional, then IMO it's not a problem. Arguably the debian-specific file is easier to maintain because at least it aims for a specific environment and can be tested there, not 'all distributions in general', which has made the file under `contrib/init/` a real pain in practice (see e.g. discussion in #12255).

ctp-tsteenholdt commented:
@laanwj That's exactly right.

laanwj commented:
Would be good to have an ACK from @TheBlueMatt here.

TheBlueMatt commented on contrib/debian/bitcoind.postrm:
```diff
@@ -0,0 +1,32 @@
+	echo "#  The bitcoin user (${BCUSER}) and data dir (${BCHOME})"
+	echo "#  were left intact."
+	echo "#"
+	echo "#  After backing up all vital data, cleanup can be completed"
```
Maybe make it explicit that they should check /var/lib/bitcoin for things like wallets?

TheBlueMatt commented on contrib/debian/bitcoind.service:
```diff
@@ -0,0 +1,45 @@
+After=network.target
+
+[Service]
+ExecStart=/usr/bin/bitcoind -daemon -conf=/etc/bitcoin/bitcoin.conf -pid=/run/bitcoind/bitcoind.pid
```
May be nice to make the datadir explicit so that we dont have /var/lib/bitcoin/.bitcoin and instead just put stuff in /var/lib/bitcoin?

TheBlueMatt commented on contrib/debian/rules:
```diff
@@ -6,18 +6,44 @@
 	rm -rf Makefile.in aclocal.m4 configure src/Makefile.in src/bitcoin-config.h.in src/build-aux src/qt/Makefile.in src/qt/test/Makefile.in src/test/Makefile.in
 
 QT=$(shell dpkg-vendor --derives-from Ubuntu && echo qt4 || echo qt5)
+# qt4 is very broken on arm
```
Can you pull the backporting of existing debian/ contents out into a separate commit to make it easier to track?

TheBlueMatt reviewed (commented):
Looks good, though I didnt test any of it.

ctp-tsteenholdt commented:
Thanks @TheBlueMatt. Each of your comments make perfect sense and I'll have them all addressed.

ctp-tsteenholdt commented:
I've addressed the suggestions made by @TheBlueMatt, run a test dpkg-buildpackage using the updated `contrib/debian/` tree and tested that the package changes work as expected, with success.

The commits have been updated to reflect the changes.

ctp-tsteenholdt commented:
Travis indicates an error but it seems unrelated to this PR?

laanwj commented:
@TheBlueMatt can you take another look?

TheBlueMatt commented:
Hmm, if you're gonna pull some of the debian changes upstream, can you just pull all of them? Especially getting the changelog and other stuff out of sync seems not ideal.

TheBlueMatt commented on contrib/debian/bitcoind.postinst:
```diff
@@ -0,0 +1,27 @@
+if [ "$1" = "configure" ]; then
+
+	# Add bitcoin user/group - this will gracefully abort if the user already exists.
+	# A homedir is never created.
```
Why not create the directory? If its empty it will be deleted on uninstall, but if ther user ever tries to use the service things will simply fail to start, no?

ctp-tsteenholdt replied:
The reason for this approach is that, if for whatever reason the user has changed ownership or permissions on the `/var/lib/bitcoin` directory, we don't want to mess with that on operations like reconfigure, update etc.

`adduser` does not let me specify custom homedir permissions, so I'm using `--no-create-home` to work around that. A missing home directory becomes a stable indicator that this is a fresh install, making it safe for me to create the directory and set the restrictive default permissions.

TheBlueMatt replied:
Oh, I misread the script, sorry.

ctp-tsteenholdt commented:
@TheBlueMatt I'm not sure what you mean by:

_Hmm, if you're gonna pull some of the debian changes upstream, can you just pull all of them? Especially getting the changelog and other stuff out of sync seems not ideal._

Upstream in this case would essentially be your PPA right? Or is there somewhere else?

I can certainly take a look at any other differences between this tree and yours and see if it makes sense to include in this PR or create a separate one for getting those in sync? Is this what you mean?

TheBlueMatt commented:
Yea, just pull the stuff from the debian PPA tar.

ctp-tsteenholdt commented:
@TheBlueMatt so I took all the changes from your upstream PPA debian tar and committed them as a single commit. After that, I committed the Systemd changes in a separate commit.

As a precaution I've run a local build and made sure that things still work as expected.

fanquake commented on contrib/debian/bitcoin-qt.desktop:
```diff
@@ -10,5 +10,4 @@ Terminal=false
 MimeType=x-scheme-handler/bitcoin;
-Categories=Office;Finance;P2P;Network;Qt;
-StartupWMClass=Bitcoin-qt
+Categories=Office;Finance;
```
Note that the additional categories were just added in #12854

ctp-tsteenholdt replied:
@fanquake thanks. I'll get that sorted out here...

Commit:
Sync contrib/debian from Matt Corallo's PPA

Commit:
Add systemd service for bitcoind

Adding systemd service for bitcoind, to provide for a simpler
out-of-the-box experience.

Configuration file is /etc/bitcoin/bitcoin.conf. This file is a
copy of the sample configuration file.

The service user 'bitcoin' is added during install. Its homedir
is in '/var/lib/bitcoin'.

bitcoind.service is disabled by default to allow the user to
configure it, before starting it the first time.

On package purge, the 'bitcoin' user as well as its homedir is
left intact, to not accidentally remove a wallet or something of
equal importance. Instead the user is presented with information
on how to perform the cleanup manually, after making sure all
important data has been backed up.

ctp-tsteenholdt commented:
So I re-resolved the conflict with #12854. This time without nuking the accepted changes of that PR. Sorry about that. Thanks @fanquake for the heads up.

laanwj commented:
@TheBlueMatt is this now ready for merge?

ctp-tsteenholdt commented:
Just tested on Ubuntu 18.04, with no issues pertaining to this PR. There is an unrelated issue with the `libdb++` version that needs to be sorted out, so I built with `--with-incompatible-bdb` to test things out.

TheBlueMatt commented on contrib/debian/rules:
```diff
@@ -6,18 +6,44 @@
+# Restart after upgrade
+override_dh_systemd_start:
+	dh_systemd_start \
+		--restart-after-upgrade
```
Why? I dont think anyone would be surprised by their bitcoind going down prior to upgrade instead of only being restarted, and this could break eg a lxc VM as you'll get a text file busy error.

ctp-tsteenholdt replied:
@TheBlueMatt, good point.

`--restart-after-upgrade` is default in `dh_systemd_start` from compat level 10, so the reason is perhaps best described as emulating modern defaults. I agree that for bitcoind, there should not be any real issues with using the defaults.

I don't quite follow the problems you mention though. As far as I can tell, an upgrade should work perfectly, even when leaving the daemon running during the upgrade. Nothing runtime related is touched during the package upgrade, only the new `bitcoind` and `bitcoin-cli` binaries are put into place, making them ready for the late restart. Unless I'm missing something, this should be no different when running in a container?

If you feel there's a risk here, I'll happily remove this!

TheBlueMatt replied:
Eh, if its default whatever, but linux generally refuses to replace binaries on disk if they are running inside of an LXC/docker container (though will happily do so during normal operation).

TheBlueMatt reviewed (commented):
Looks good modulo the one restart question I had.

TheBlueMatt commented:
utACK 2a87b1b07c5c4f8b9b34747c5f254c2ae1e824bf with or without restart-after-upgrade.

luke-jr commented:
How will this affect systems that don't have/use systemd?

ctp-tsteenholdt commented:
@luke-jr this affects only debian-based systems (including Ubuntu) with systemd.

ctp-tsteenholdt commented:
Older debian systems that are missing the dh-systemd package, will fail to build the packages, yes. I believe the (supported) Debian/Ubuntu versions that are affected by this, are Debian 7 (LTS version is EOL at the end of May 2018) and Ubuntu 14.04 LTS (EOL in April 2019).

I'm running a test on Ubuntu 14.04 right now, to see what kind of issues we're running into on that platform.

ctp-tsteenholdt commented:
As it turns out, Ubuntu 14.04 LTS actually does have a `dh-systemd` package and apart from the `libdb4.8++-dev` build dependency, that was also causing problems on Ubuntu 18.04, the build is working just fine.

Since the distro does not actually use systemd, the included `bitcoind.service` simply does nothing.

luke-jr commented:
I'm less concerned with merely old distros, as much as ones that have simply rejected the mandatory-systemd nonsense, such as Devuan.

ctp-tsteenholdt commented:
@luke-jr to build the package, you'd need to have `dh-systemd` installed, but you do *not* need `systemd` to install or use it. If you don't have `systemd` on the system, you simply can't use the systemd service configuration that's included here.

ctp-tsteenholdt commented:
Did that last comment answer your questions?

laanwj commented:
I'm going to merge this as @TheBlueMatt utACKed this - he's the ppa maintainer. 
I don't think other distributions are relevant here as this concerns explicitly the `debian` directory.
Other improvements can be made later.

TheBlueMatt commented:
Err, no, Luke's point is good and would break several PPA-supported releases.

On May 1, 2018 2:52:47 PM UTC, "Wladimir J. van der Laan" <notifications@github.com> wrote:
>I'm going to merge this as @TheBlueMatt utACKed this - he's the ppa
>maintainer. 
>I don't think other distributions are relevant here as this concerns
>explicitly the `debian` directory.
>Other improvements can be made later.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/12769#issuecomment-385690313

TheBlueMatt commented:
Oh, hmm, I appears to be wrong, it looks like the dh-systemd package *is* on trusty, I suppose its fine, then.

laanwj commented:
Had to revert this, please file again in 2019 (or whenever relevant).

ctp-tsteenholdt commented:
I believe dh-systemd is indeed available for all relevant distributions - It certainly is for Devuan, Ubuntu Trusty (and later) and Debian 7 (and later).

laanwj commented:
We're switching to a different strategy, distribution packaging files will be under this repository from now on:
https://github.com/bitcoin-core/packaging

ctp-tsteenholdt commented:
Okay... Does this mean I should rework the changes to go in the new path then?

MarcoFalke commented:
Indeed, just cherry-pick the commits and submit them to the other repo.
